packages/medusa/src/services/idempotency-key.ts

```diff
@@ -0,0 +1,197 @@
+import { MedusaError } from "medusa-core-utils"
+import { v4 } from "uuid"
+import { TransactionBaseService } from "../interfaces"
+import { DeepPartial, EntityManager } from "typeorm"
+import { IdempotencyKeyRepository } from "../repositories/idempotency-key"
+import { IdempotencyKey } from "../models"
+import { CreateIdempotencyKeyInput } from "../types/idempotency-key"
+
+const KEY_LOCKED_TIMEOUT = 1000
+
+type InjectedDependencies = {
+  manager: EntityManager
+  idempotencyKeyRepository: typeof IdempotencyKeyRepository
+}
+
+class IdempotencyKeyService extends TransactionBaseService<IdempotencyKeyService> {
+  protected manager_: EntityManager
+  protected transactionManager_: EntityManager | undefined
+
+  protected readonly idempotencyKeyRepository_: typeof IdempotencyKeyRepository
+
+  constructor({ manager, idempotencyKeyRepository }: InjectedDependencies) {
+    super({ manager, idempotencyKeyRepository })
+
+    this.manager_ = manager
+    this.idempotencyKeyRepository_ = idempotencyKeyRepository
+  }
+
+  /**
+   * Execute the initial steps in a idempotent request.
+   * @param headerKey - potential idempotency key from header
+   * @param reqMethod - method of request
+   * @param reqParams - params of request
+   * @param reqPath - path of request
+   * @return the existing or created idempotency key
+   */
+  async initializeRequest(
+    headerKey: string,
+    reqMethod: string,
+    reqParams: Record<string, unknown>,
+    reqPath: string
+  ): Promise<IdempotencyKey> {
+    return await this.atomicPhase_(async () => {
+      return (
+        (await this.retrieve(headerKey).catch(() => void 0)) ??
+        (await this.create({
+          request_method: reqMethod,
+          request_params: reqParams,
+          request_path: reqPath,
+        }))
```

Comment:
thought: I believe we are compromising readability just to use the coalescing operator here. Would rather have a method body easy to comprehend at first glance. What do you think?